scala/private/phases/phase_dependency.bzl

```diff
@@ -0,0 +1,49 @@
+# Gathers information about dependency mode and analysis
+
+load(
+    "@io_bazel_rules_scala//scala/private:dependency.bzl",
+    "get_dependency_mode",
+    "get_strict_deps_mode",
+    "get_unused_deps_mode",
+    "new_dependency_info",
+)
+
+def _phase_dependency_default(ctx, p, args = struct()):
+    return _phase_dependency(
+        ctx = ctx,
+        p = p,
+        unused_deps_always_off = args.unused_deps_always_off if hasattr(args, "unused_deps_always_off") else False,
+        strict_deps_always_off = args.strict_deps_always_off if hasattr(args, "strict_deps_always_off") else False,
+    )
+
+def _phase_dependency(
```

Comment:
If we have a dedicated phase for dependency, is it possible to move all the logic in `//scala/private:dependency.bzl` to this phase? This way we can really have a centralized dependency logic.

I can see a reason to have a shared file since some of the functions are reused in other places. My vision (as the author of phase architecture) is to put code in shared files as less as possible. Hopefully we will get there when we clean up `rule_impls.bzl` and `common.bzl`.

WDYT?

Reply:
Okay there are 3 places not the phase where we import dependency.bzl

- scalapb_aspect.bzl - does this actually participate in phases? If not then don't think we would be able to do this. If it does then I probably need to reread but it should be possible.
- twitter_scrooge.bzl - same thing as ^
- phase_compile.bzl - This looks probably doable - since we are just looking at the files in the rule and seeing if any end in `.java`. This does mean duplicating the logic of seeing which files are java files, or else having some place where we put that logic - where would be a good place?

Reply:
There is indeed a follow-up task for moving remaining rules to phase architecture #938, `scalapb_aspect.bzl` and `twitter_scrooge.bzl` may be included. But at a brief glance, I don't think they fit into existing phases.

Probably a bad idea: generate `phase_dependency_scalapb_aspect` and `phase_dependency_twitter_scrooge` and use these functions as normal. This way we are making more tech-debt but we can avoid having `//scala/private:dependency.bzl`.

We should wait for @ittaiz on this before we decide but I would like to know WDYT?

Reply:
If we just say right now `dependency.bzl` is the source of truth but later `phase_dependency.bzl` becomes the source of truth when some other stuff is sorted out that seems the straightforward way to me.

(Then we just need to sort out `phase_compile.bzl` - do you have suggestions on how to break out the java file separation logic?)

But if people are in favor of making `phase_dependency_scalapb_aspect` etc and the somewhat strange imports which ensure, I am fine with that.

Reply:
> do you have suggestions on how to break out the java file separation logic?

Yes, I do think we should avoid duplication and have a function that `phase_dependency` can use to set `unused_deps_mode = "off"`. Then `phase_compile` doesn't need `new_dependency_info`. WDYT?

I can work on this.

Reply:
Ok - so `phase_dependency` and `phase_compile` will call the function (which I guess lives in `rule_impls.bzl`) something like `get_java_sources()`, and `phase_dependency` will turn unused deps off whenever `get_java_sources()` is nonempty. Is that correct?

And in the meantime are we keeping around `dependency.bzl` or doing the thing of having `phase_dependency_scalapb_aspect` in `dependency_phase.bzl`?

Reply:
> Ok - so phase_dependency and phase_compile will call the function (which I guess lives in rule_impls.bzl) something like get_java_sources(), and phase_dependency will turn unused deps off whenever get_java_sources() is nonempty. Is that correct?

That is correct. I believe it will be better this way.

> And in the meantime are we keeping around dependency.bzl or doing the thing of having phase_dependency_scalapb_aspect in dependency_phase.bzl?

We have to wait for maintainers. Sorry, I am not one, but I would like to help.

Reply:
How does this PR look? Do you think it will get us what we want? #979

Reply:
With that PR should we also have a constant for java extension? But otherwise seems like it does what we want.

I will make updates after maintainers look over the discussions.

Reply:
> With that PR should we also have a constant for java extension?

Good idea. Updated.

Reply:
FYI, the PR has been merged. The java extension is in `scala/private/paths.bzl`.

Reply:
Fixed the part with calling from phase_compile.scala, will await @ittaiz judgement for whether to do `phase_dependency_scalapb_aspect` now or later

Reply:
I'd also like for us to remove shared files but like @borkaehw says we need to finish move remaining rules. 
I'm against the synthetic phase for the aspects and I think that the right trade off for now is:
1. Have phases only use data from other phases as much as possible.
2. Where phases must use functions because non phase rules/aspects need them those functions should be in distinct bzl files with clear roles. This means eventually killing rule_impls.bzl and common.bzl in favor of smaller files.

Reply:
I believe 1. is the case, as the only place which uses `dependency.yml` is `phase_dependency.yml` and actually all other phases are able to get info from `phase_dependency`. As for 2. I think `dependency.yml` also serves that purpose. So given the fact we don't want synthetic phases I don't think we need any changes here.

Reply:
You mean that if a function is currently in `dependency.bzl` but doesn't need to be, then move it to `phase_dependency.bzl`? It seems like it splits logic between two places, but given that it is in short term plans to move everything to `phase_dependency.bzl` anyways am fine with such, so will do so.

Reply:
Yes. The idea is for code in the shared files to stick out like a sore thumb.